api/books.go

```diff
@@ -0,0 +1,114 @@
+package api
+
+import (
+	"encoding/json"
+	"github.com/PutskouDzmitry/golang-training-Library/pkg/data"
+	"gorilla/mux"
+	"log"
+	"net/http"
+	"strconv"
+)
+
+type bookAPI struct {
+	data *data.BookData
+}
+
+func ServeUserResource(r *mux.Router, data data.BookData) {
+	api := &bookAPI{data: &data}
+	r.HandleFunc("/books", api.getAllBooks).Methods("GET")
+	r.HandleFunc("/book{id}", api.getOneBook).Methods("GET")
+	r.HandleFunc("/books", api.createBook).Methods("POST")
+	r.HandleFunc("/books{id}/{number}", api.updateBook).Methods("PUT")
+	r.HandleFunc("/books{id}", api.deleteBook).Methods("DELETE")
+}
+
+func (a bookAPI) getAllBooks(writer http.ResponseWriter, request *http.Request) {
+	users, err := a.data.ReadAll()
+	if err != nil {
+		_, err := writer.Write([]byte("got an error when tried to get users"))
+		if err != nil {
+			log.Println(err)
+		}
+	}
+	err = json.NewEncoder(writer).Encode(users)
+	if err != nil {
+		log.Println(err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+}
+
+func (a bookAPI) getOneBook(writer http.ResponseWriter, request *http.Request) {
+	idRequest := mux.Vars(request)
+	id, err := strconv.Atoi(idRequest["id"])
+	user, err := a.data.Read(id)
+	if err != nil {
+		_, err := writer.Write([]byte("got an error when tried to get users"))
+		if err != nil {
+			log.Println(err)
```

Comment:
You just printing an error, and not exit from function, so in step: encoding go model to json you will get an error again.
So please add writer.WriteHeader(http.StatusInternalServerError) and exit of function after printing log message.